### PR TITLE
Add functionality for simulating greedy pricing

### PR DESCRIPTION
**Pull request by jiangdaniel**

### CODE REVIEW

File: learner_utils.py [hunk 1]
```diff
@@ -28,6 +28,22 @@ def learner_decisions(subpops, current = None, min_fn=quadratic_min):
             new_thetas.append(theta_i)
     return new_thetas
 
+def learner_max_price(thetas, prices, subpops, min_fn=quadratic_min):
```
Comment (jiangdaniel) [inline learner_utils.py:31]: I'll add documentation

File: learner_utils.py [hunk 2]
```diff
@@ -28,6 +28,22 @@ def learner_decisions(subpops, current = None, min_fn=quadratic_min):
...
+    n_learners = len(thetas)
+    new_prices = [0] * len(prices)
+    #price_candidates = np.linspace(0, 100, 51)
```
Comment (jiangdaniel) [inline learner_utils.py:34]: will remove

File: learner_utils.py [hunk 3]
```diff
@@ -28,6 +28,22 @@ def learner_decisions(subpops, current = None, min_fn=quadratic_min):
...
+    for i in range(n_learners):
+        price_candidates = [prices[i] + 1, prices[i] * 2, prices[i] / 2]
```
Comment (jiangdaniel) [inline learner_utils.py:36]: will add the identity price `prices[i]` as an option (I didn't realize I had forgotten to do that before)

File: learner_utils.py [hunk 4]
```diff
@@ -28,6 +28,22 @@ def learner_decisions(subpops, current = None, min_fn=quadratic_min):
...
+        profit_candidates = []
+        for p in price_candidates:
+            tmp = prices[i]
+            prices[i] = p
+            new_alphas = subpop_decisions(thetas, subpops, epsilon=0.1, prices=prices, dryrun=True)
+            profit_candidates.append(new_alphas[:, i].sum() * p)  # warning: assumes beta is uniform
```
Comment (jiangdaniel) [inline learner_utils.py:42]: will fix this calculation to account for non-uniform beta

File: subpop.py [hunk 5]
```diff
@@ -24,7 +24,7 @@ def __init__(self, phi, beta, alphas, cov=None, sigmasq=1, price_sensitivity=0):
         self.converged = False
         self.converged_for = 0
     
-    def update_alpha(self, thetas, epsilon=0.1, prices = None):
+    def update_alpha(self, thetas, epsilon=0.1, prices = None, dryrun=False):
```
Comment (jiangdaniel) [inline subpop.py:27]: will add documentation for new argument

File: learner_utils.py (hunk 4 above)
Comment (mcurmei627) [reply learner_utils.py:42]: maybe make a comment in the code 
#TODO

File: subpop.py [hunk 6]
```diff
@@ -38,11 +38,13 @@ def update_alpha(self, thetas, epsilon=0.1, prices = None):
...
-            self.converged_for += 1
-        self.alphas = new_alphas
-        self.t += 1 
+        if not dryrun:
+            self.converged = np.allclose(new_alphas, self.alphas)
+            if self.converged:
+                self.converged_for += 1
+            self.alphas = new_alphas
+            self.t += 1
+        return new_alphas
```
Comment (mcurmei627) [inline subpop.py:47]: nice

**Review by mcurmei627 — APPROVED** [review]
Looks nice! Ty

**Pull request merged**